File: package/main/src/Object/flattenObject.ts
import { isPlainObject } from "./isPlainObject";

/**
 * Recursively flattens a nested plain object into a single-level object
 * keyed by joined paths. Arrays and non-plain objects are kept as-is.
 *
 * @param {Record<string, unknown>} object - Source nested object
 * @param {string} [separator="."] - Separator used to join path segments
 * @returns {Record<string, unknown>} A flat object
 * @example
 * flattenObject({ a: { b: { c: 1 } }, d: 2 }); // { "a.b.c": 1, d: 2 }
 */
export const flattenObject = <T extends Record<string, unknown>>(
  object: T,
  separator = ".",
): Record<string, unknown> => {
  const result: Record<string, unknown> = {};

  const walk = (current: Record<string, unknown>, prefix: string): void => {
    for (const key of Object.keys(current)) {
      const value = current[key];
      const nextKey = prefix ? `${prefix}${separator}${key}` : key;
      if (isPlainObject(value) && Object.keys(value).length > 0) {
        walk(value, nextKey);
      } else {
        result[nextKey] = value;
      }
    }
  };

  walk(object, "");
  return result;
};

File: package/umt_python/src/object/flatten_object.py
from typing import Any

from .object_is_plain import object_is_plain


def flatten_object(
    object_: dict[str, Any],
    separator: str = ".",
) -> dict[str, Any]:
    """
    Recursively flattens a nested plain object into a single-level object
    keyed by joined paths. Arrays (lists) and non-plain objects are kept as-is.

    Empty plain objects are treated as leaf values rather than being removed.

    Args:
        object_ (dict[str, Any]): Source nested object.
        separator (str): Separator used to join path segments. Defaults to ".".

    Returns:
        dict[str, Any]: A flat object keyed by joined paths.

    Example:
        >>> flatten_object({"a": {"b": {"c": 1}}, "d": 2})
        {'a.b.c': 1, 'd': 2}
        >>> flatten_object({"a": {"b": 1}}, "/")
        {'a/b': 1}
        >>> flatten_object({"a": {}})
        {'a': {}}
    """
    result: dict[str, Any] = {}

    def walk(current: dict[str, Any], prefix: str) -> None:
        for key, value in current.items():
            next_key = f"{prefix}{separator}{key}" if prefix else key
            if object_is_plain(value) and len(value) > 0:
                walk(value, next_key)
            else:
                result[next_key] = value

    walk(object_, "")
    return result
